osu.Game.Rulesets.Osu/UI/OsuDrawableTouchInputHandler.cs

```diff
@@ -0,0 +1,109 @@
+using osu.Game.Screens.Play;
+
+namespace osu.Game.Rulesets.Osu.UI
+{
+    public class OsuDrawableTouchInputHandler : Drawable
+    {
+        public const TouchSource CURSOR_TOUCH = TouchSource.Touch1;
+
+        /// <summary>
+        /// How many taps (taps referring as streaming touch input) can be registered.
+        /// </summary>
+        private const int tap_touches_limit = 2;
+
+        /// <summary>
+        /// How many concurrent touches can be registered.
+        /// </summary>
+        private const int concurrent_touches_limit = tap_touches_limit + 1;
+
+        /// <summary>
+        /// The index for the last concurrent able touch.
+        /// </summary>
+        private const int last_concurrent_touch_index = concurrent_touches_limit - 1;
+
+        public readonly HashSet<TouchSource> AllowedTouchSources = Enum.GetValues(typeof(TouchSource)).Cast<TouchSource>().Take(concurrent_touches_limit).ToHashSet();
+
+        private readonly Dictionary<TouchSource, OsuAction> touchActions = new Dictionary<TouchSource, OsuAction>();
+
+        private readonly OsuInputManager osuInputManager;
+
+        [Resolved(canBeNull: true)]
+        private Player player { get; set; }
+
+        private bool detectedTouchscreen;
+
+        private readonly BindableInt detectedTouches = new BindableInt
+        {
+            MaxValue = 10
+        };
+
+        private int getTouchIndex(TouchSource source) => source - TouchSource.Touch1;
+
+        public OsuDrawableTouchInputHandler(OsuInputManager inputManager)
+        {
+            osuInputManager = inputManager;
+            foreach (var source in AllowedTouchSources)
+                touchActions.Add(source, getTouchIndex(source) % 2 == 0 ? OsuAction.LeftButton : OsuAction.RightButton);
+        }
+
+        private bool isTapTouch(TouchSource source) => source != CURSOR_TOUCH || !osuInputManager.AllowUserCursorMovement;
+
+        private bool isCursorTouch(TouchSource source) => !isTapTouch(source);
+
+        private bool isValidTouchInput(int index) => index <= last_concurrent_touch_index;
+
+        protected override bool OnTouchDown(TouchDownEvent e)
+        {
+            var source = e.Touch.Source;
+            int sourceIndex = getTouchIndex(source);
+
+            if (!isValidTouchInput(sourceIndex))
+                return false;
+
+            osuInputManager.DragMode = sourceIndex == last_concurrent_touch_index;
+
+            if (isCursorTouch(source))
+            {
+                if (!detectedTouchscreen && ++detectedTouches.Value == detectedTouches.MaxValue)
+                {
+                    detectedTouchscreen = true;
+                    player.Score.ScoreInfo.Mods = player.Score.ScoreInfo.Mods.Append(new OsuModTouchDevice()).ToArray();
+                }
+
+                return base.OnTouchDown(e);
+            }
+
+            osuInputManager.KeyBindingContainer.TriggerPressed(touchActions[source]);
+
+            return true;
+        }
```

Comment:
I don't think this whole complexity is required. At the point where `OnTouchDown()` is called, I would expect the device to be guaranteed to be a touch device.

The complexity is in detecting if a touch device is acting as a mouse device, as I believe is currently the case for windows touch screens, and that detecting would involve `OnMouseDown()` instead.

Reply:
Although that, i still think it should only detect the play as touch device if you tap with your touch screen a few times because touch device is a mod that rewards less pp, this leave a small "grace period" for players who play touch but dragging instead, for the windows part, yeah i didn't knew about that

Reply:
> The complexity is in detecting if a touch device is acting as a mouse device, as I believe is currently the case for windows touch screens, and that detecting would involve `OnMouseDown()` instead.

I fixed this part in my latest commit.